File: problems/2373/Lc2373.cpp
#include <vector>
using namespace std;

class Solution {
public:
  int max9(int arg0, int arg1, int arg2, int arg3, int arg4, int arg5, int arg6,
           int arg7, int arg8) {
    if (arg0 < arg1) arg0 = arg1;
    if (arg0 < arg2) arg0 = arg2;
    if (arg0 < arg3) arg0 = arg3;
    if (arg0 < arg4) arg0 = arg4;
    if (arg0 < arg5) arg0 = arg5;
    if (arg0 < arg6) arg0 = arg6;
    if (arg0 < arg7) arg0 = arg7;
    if (arg0 < arg8) arg0 = arg8;
    return arg0;
  }
  vector<vector<int>> largestLocal(vector<vector<int>>& grid) {
    size_t bound = grid.size() - 2;
    vector<vector<int>> result(bound);
    for (size_t rpos = 0; rpos < bound; ++rpos) {
      for (size_t cpos = 0; cpos < bound; ++cpos) {
        result[rpos].push_back(
            max9(grid[rpos][cpos], grid[rpos][cpos + 1], grid[rpos][cpos + 2],
                 grid[rpos + 1][cpos], grid[rpos + 1][cpos + 1],
                 grid[rpos + 1][cpos + 2], grid[rpos + 2][cpos],
                 grid[rpos + 2][cpos + 1], grid[rpos + 2][cpos + 2]));
      }
    }
    return result;
  }
};

int main() {
  return 0;
}

File: problems/2373/Lc2373.ts
import { nestedArraysEqual } from '../common/Common';

function max(...args: number[]): number {
  let result = args[0];
  for (let pos = 1, bound = args.length; pos < bound; pos++) {
    if (result < args[pos]) result = args[pos];
  }
  return result;
};

function largestLocal(grid: number[][]): number[][] {
  const n = grid.length;
  const result: number[][] = [];
  const bound = n - 2;
  for (let rpos = 0; rpos < bound; rpos++) {
    result.push([]);
    for (let cpos = 0; cpos < bound; cpos++) {
      result[rpos].push(max(
        grid[rpos][cpos], grid[rpos][cpos + 1], grid[rpos][cpos + 2],
        grid[rpos + 1][cpos], grid[rpos + 1][cpos + 1], grid[rpos + 1][cpos + 2],
        grid[rpos + 2][cpos], grid[rpos + 2][cpos + 1], grid[rpos + 2][cpos + 2],
      ));
    }
  }
  return result;
};

function main() {
  let result: number[][], expected: number[][], grid: number[][];

  grid = [
    [9, 9, 8, 1],
    [5, 6, 2, 6],
    [8, 2, 6, 4],
    [6, 2, 2, 2]]
  result = largestLocal(grid);
  expected = [
    [9, 9],
    [8, 6]];
  if (!nestedArraysEqual(expected, result)) new Error(`expected ${expected}, but ${result}`);

  grid = [
    [9, 9, 8, 1],
    [5, 6, 2, 6],
    [8, 2, 6, 4],
    [6, 2, 2, 2]];
  result = largestLocal(grid);
  expected = [
    [2, 2, 2],
    [2, 2, 2],
    [2, 2, 2]];
  if (!nestedArraysEqual(expected, result)) new Error(`expected ${expected}, but ${result}`);
};
main();
